### PR TITLE
Issue attach vm to volume

### PR DESCRIPTION
**Pull request by bharatE2E**

### CODE REVIEW

File: e2e/blockstorage/resource_block_storage.go [hunk 1]
```diff
@@ -155,7 +169,12 @@ func resourceUpdateBlockStorage(ctx context.Context, d *schema.ResourceData, m i
 	blockStorageID := d.Id()
 	project_id := d.Get("project_id").(int)
 	location := d.Get("location").(string)
+	status := d.Get("status").(string)
 
+	if status == "ERROR" {
+		rollbackChanges(d)
+		return diag.Errorf("Block Storage is in ERROR state. Cannot be updated. please destroy and recreate")
```
Comment (tarune2e) [inline e2e/blockstorage/resource_block_storage.go]: please change the error message we can not ask the customer to delete and recreate the resource

File: e2e/blockstorage/resource_block_storage.go [hunk 2]
```diff
@@ -346,15 +368,20 @@ func waitForDetach(apiClient *client.Client, blockStorageID string, project_id i
 		if data["status"] == "Available" {
 			break
 		}
-		// Wait for 2 seconds before checking the status again (is Volume Detached?)
-		time.Sleep(2 * time.Second)
+		// Wait for some time before checking the status again (is Volume Detached?)
```
Comment (tarune2e) [inline e2e/blockstorage/resource_block_storage.go]: remove this code if not required

File: e2e/blockstorage/resource_block_storage.go [hunk 3]
```diff
@@ -346,15 +368,20 @@ func waitForDetach(apiClient *client.Client, blockStorageID string, project_id i
...
 }
 
 func validateSize(d *schema.ResourceData, m interface{}) error {
 	apiClient := m.(*client.Client)
 
+	_, err := apiClient.GetAllProjects(d.Get("project_id").(int), d.Get("location").(string))
+	if err != nil {
+		return err
+	}
+
```
Comment (tarune2e) [inline e2e/blockstorage/resource_block_storage.go]: remvoe this api call. No need to make extra api call

File: e2e/blockstorage/resource_block_storage.go (hunk 1 above)
Comment (bharatE2E) [reply e2e/blockstorage/resource_block_storage.go]: done

File: e2e/blockstorage/resource_block_storage.go (hunk 2 above)
Comment (bharatE2E) [reply e2e/blockstorage/resource_block_storage.go]: done

File: e2e/blockstorage/resource_block_storage.go (hunk 3 above)
Comment (bharatE2E) [reply e2e/blockstorage/resource_block_storage.go]: done

File: client/block_storage_client.go [hunk 4]
```diff
@@ -232,3 +233,34 @@ func CheckResponseStatusForBlock(response *http.Response) error {
 	}
 	return nil
 }
+
+func (c *Client) GetAllProjects(project_id int, location string) (map[string]interface{}, error) {
```
Comment (ParitoshThapliyal59) [inline client/block_storage_client.go]: Let's remove these lines, if we are not planning to use this function in future.

File: e2e/node/resource_node.go [hunk 5]
```diff
@@ -301,7 +327,7 @@ func resourceCreateNode(ctx context.Context, d *schema.ResourceData, m interface
 		Enable_bitninja:   d.Get("enable_bitninja").(bool),
 		Is_ipv6_availed:   d.Get("is_ipv6_availed").(bool),
 		Is_saved_image:    d.Get("is_saved_image").(bool),
-		Region:            d.Get("region").(string),
+		// Region:            d.Get("location").(string),
```
Comment (ParitoshThapliyal59) [inline e2e/node/resource_node.go]: Let's remove the commented line. Unless kept intentionally for some reason.

File: e2e/node/resource_node.go [hunk 6]
```diff
@@ -410,11 +437,14 @@ func resourceUpdateNode(ctx context.Context, d *schema.ResourceData, m interface
 	nodeId := d.Id()
 	project_id := d.Get("project_id").(string)
 	location := d.Get("location").(string)
+	status := d.Get("status").(string)
+	if status == "Failed" {
```
Comment (ParitoshThapliyal59) [inline e2e/node/resource_node.go]: the value "Failed" must be defined in a constant. If we have pre-defined states, then those should be defined in a constant. All such states must be kept in a single array or hashmap and fetched from there.

File: e2e/node/resource_node.go [hunk 7]
```diff
@@ -428,6 +458,8 @@ func resourceUpdateNode(ctx context.Context, d *schema.ResourceData, m interface
 	if d.HasChange("power_status") {
 		nodestatus := d.Get("status").(string)
 		if nodestatus == "Creating" || nodestatus == "Reinstalling" {
```
Comment (ParitoshThapliyal59) [inline e2e/node/resource_node.go]: "Creating" and "Reinstalling" is the same case as "Failed" state. Put these in a constant.

File: e2e/node/resource_node.go [hunk 8]
```diff
@@ -641,32 +673,58 @@ func resourceUpdateNode(ctx context.Context, d *schema.ResourceData, m interface
 			}
 		}
 
-		prevBlockID, currBlockID := d.GetChange("block_storage_id")
-		log.Printf("[INFO] prevID %v, currID %v", prevBlockID, currBlockID)
+		prevBlockIDArray, currBlockIDArray := d.GetChange("block_storage_ids")
+
+		if d.Get("plan").(string)[0:2] == "C2" {
```
Comment (ParitoshThapliyal59) [inline e2e/node/resource_node.go]: Do not use hard-coded values. Put "C2" in constant.

File: e2e/node/resource_node.go [hunk 9]
```diff
@@ -723,16 +778,17 @@ func waitForPoweringOffOn(m interface{}, nodeId string, project_id string) error
...
 
 		nodeInfo, err := apiClient.GetNode(nodeId, project_id)
 		if err != nil {
 			log.Printf("[ERROR] Error getting Node Info inside Plan Upgrade. Error : %s", err)
 			return err
 		}
 		data := nodeInfo["data"].(map[string]interface{})
-		if !(data["status"] == "Powering on" || data["status"] == "Powering off") {
+		log.Printf("[INFO] Node Status : %s", data["status"])
+		if data["status"] == "Running" || data["status"] == "Powered off" {
```
Comment (ParitoshThapliyal59) [inline e2e/node/resource_node.go]: Put VM states in constant file.

File: e2e/node/resource_node.go [hunk 10]
```diff
@@ -755,3 +811,169 @@ func getDefaultSG(response map[string]interface{}) int {
...
+		time.Sleep(constants.WAIT_TIMEOUT * time.Second)
+
+		response, err := apiClient.CheckNodeLCMState(nodeId, project_id, location)
+		if err != nil {
+			log.Printf("[ERROR] Error getting lcm_state %s", err)
+			return diag.FromErr(err)
+		}
+		data := response["data"].(map[string]interface{})
+		log.Printf("[INFO] waitForDesiredState data : %+v", data)
+		if !(data["lcm_state"].(string) == "Hotplug" || data["lcm_state"].(string) == "HOTPLUG_PROLOG_POWEROFF" || data["lcm_state"].(string) == "HOTPLUG_EPILOG_POWEROFF") {
```
Comment (ParitoshThapliyal59) [inline e2e/node/resource_node.go]: Need these hard-coded values in Constant.

**Pull request merged**